### PR TITLE
Modify SymmetricCipher overlaps buffer checks

### LINKED ISSUE #292: Fix unary oparator in the SymmetricCipher
These two lines have the incorrect unary operator `=`.

That needs to be `+` --> `(outputOffset < (inputOffset + inputLen))`

https://github.com/IBM/OpenJCEPlus/blob/0464ecbf69b0b908f77372af37cdce40b550ed9f/src/main/java/com/ibm/crypto/plus/provider/ock/SymmetricCipher.java#L374

https://github.com/IBM/OpenJCEPlus/blob/0464ecbf69b0b908f77372af37cdce40b550ed9f/src/main/java/com/ibm/crypto/plus/provider/ock/SymmetricCipher.java#L504

### PR DESCRIPTION
**Pull request by jasonkatonica**
The `doFinal` and `update` methods in the SymmetricCipher class have logic to detect if a input buffer and output buffer are overlapping in any sort of way. If they are then a copy of the input is made to a separate location to ensure a safe operation can then take place on the data.

There is a logic problem where the calculation of overlap was incorrect and used a `=` instead of a `+` to calculate if the input and output overlapped. The comment above this code block was correct and remains unmodified.

The test `BaseTestAESGCMCopySafe` was modified to test this condition which it was attempting to do in the past however it was not operating upon input and output buffers that had the same memory address. The test was modified to exercise various overlapping conditions for AES GCM and also AES CBC accordingly. The test was then renamed to `BaseTestAESCopySafe` since we are now testing both GCM and CBC modes.

Fixes #292

Signed-off-by: Jason Katonica <katonica@us.ibm.com>

### CODE REVIEW

File: src/test/java/ibm/jceplus/junit/openjceplus/TestAESCopySafe.java [hunk 1]
```diff
@@ -38,7 +38,7 @@ public static void main(String[] args) throws Exception {
     //
     //
     public static Test suite() {
-        TestSuite suite = new TestSuite(TestAESGCMCopySafe.class);
+        TestSuite suite = new TestSuite(TestAESCopySafe.class);
```
Comment (KostasTsiounis) [inline src/test/java/ibm/jceplus/junit/openjceplus/TestAESCopySafe.java:41]: Should we remove this method and `main` as well?

File: src/test/java/ibm/jceplus/junit/openjceplusfips/TestAESCopySafe.java [hunk 2]
```diff
@@ -39,7 +39,7 @@ public static void main(String[] args) throws Exception {
     //
     //
     public static Test suite() {
-        TestSuite suite = new TestSuite(TestAESGCMCopySafe.class);
+        TestSuite suite = new TestSuite(TestAESCopySafe.class);
```
Comment (KostasTsiounis) [inline src/test/java/ibm/jceplus/junit/openjceplusfips/TestAESCopySafe.java:42]: Should we remove this method and `main` as well?

File: src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                    if (mode == MODE.CBC) {
+                        System.out.println("Testing update");
+                        doTest(c, params, inputOffset, outputOffset, mode, true);
+                    }
+                }
+            }
+        }
+    }
+
+    private void doTest(Cipher c, AlgorithmParameterSpec params, int inputOffset, int outputOffset, MODE mode, boolean isUpdate)
```
Comment (KostasTsiounis) [inline src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java:71]: The `params` parameter doesn't seem to be used any more. I think it should;d be removed.

File: src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            System.out.println ("cipherText.length: " + cipherText.length);
+        }
+
+        // A GCM cipher must be reinitialized since the IV is typically not supposed to be reused. This
+        // test however expects a deterministic encrypted value in order to evaluate if the operation
+        // worked as expected.
+        initCipher(c, mode, true);
+
+        System.arraycopy(clearText, 0, workingBuffer, inputOffset, INPUT_LENGTH);
+        if (DEBUG) {
```
Comment (KostasTsiounis) [inline src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java:102]: Should we specify that we are testing encryption in this step?

File: src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        // Test decryption now, we should get back the original clear text value as a result.
+        if (DEBUG) {
+            System.out.println("c.init(DECRYPT) with KEY and Params");
+        }
+        initCipher(c, mode, false);
+
+        Arrays.fill(workingBuffer, 0, INPUT_LENGTH, (byte)0x00);
+        System.arraycopy(cipherText, 0, workingBuffer, inputOffset, cipherText.length);
+
+        if (DEBUG) {
```
Comment (KostasTsiounis) [inline src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java:132]: Should we specify that we are testing decryption in this step?

File: src/test/java/ibm/jceplus/junit/openjceplus/TestAESCopySafe.java (hunk 1 above)
Comment (jasonkatonica) [reply src/test/java/ibm/jceplus/junit/openjceplus/TestAESCopySafe.java:41]: As part of the work done in https://github.com/IBM/OpenJCEPlus/pull/252 I did remove all `main` methods and migrated each test to Junit 5. Once this PR has been approved and merged is it OK that i do that work in that PR? I will need to resolve differences anyway in order to update the Junit 5 PR for the test name change being made here also.

File: src/test/java/ibm/jceplus/junit/openjceplusfips/TestAESCopySafe.java (hunk 2 above)
Comment (jasonkatonica) [reply src/test/java/ibm/jceplus/junit/openjceplusfips/TestAESCopySafe.java:42]: Same comment/question as above.

File: src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java (hunk 3 above)
Comment (jasonkatonica) [reply src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java]: Yes agreed i did move `params` into the helper method `initCipher` here so it is no longer needed to be passed around. Adjusted this in new commit https://github.com/IBM/OpenJCEPlus/pull/295/commits/46a71f1f5e9d53cc94e3336b27fbf231deaeeab3 .

File: src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java (hunk 4 above)
Comment (jasonkatonica) [reply src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java:101]: Yes its prob best to state we are doing encryption or decryption in this DEBUG statement.  Adjusted this in new commit https://github.com/IBM/OpenJCEPlus/pull/295/commits/46a71f1f5e9d53cc94e3336b27fbf231deaeeab3 .

File: src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java (hunk 5 above)
Comment (jasonkatonica) [reply src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java:132]: Yes its prob best to state we are doing encryption or decryption in this DEBUG statement.  Adjusted this in new commit https://github.com/IBM/OpenJCEPlus/pull/295/commits/46a71f1f5e9d53cc94e3336b27fbf231deaeeab3 .

**Review by KostasTsiounis — APPROVED** [review]
LGTM

**Review by taoliult — APPROVED** [review]
LGTM

File: src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+        byte[] IV = new byte[16];
+
+        if (mode == MODE.GCM) {
+            AlgorithmParameterSpec params = new GCMParameterSpec(specifiedKeySize, IV);
+            // Re-initialize with only key value first to bypass the
+            // key+IV uniqueness check for GCM encryption.
+            if (isEncrypt) {
+                c.init(Cipher.ENCRYPT_MODE, KEY);
+                c.init(Cipher.ENCRYPT_MODE, KEY, params);
```
Comment (JinhangZhang) [inline src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java:168]: should we just use the second `init` function to avoid a second initialization?

**Review by johnpeck-us-ibm — APPROVED** [review]

**Comment by jasonkatonica** [comment]
x86_64_linux,ppc64le_linux,s390x_linux,x86_64_windows

**Pull request merged**